### PR TITLE
Add a vendordep gradle task

### LINKED ISSUE #417: Implement a vendordep task for adding deps from the commandline
I can try to implement this myself I just need to know where in the repo to put the files, if this is a wanted feature. 
The use/syntax should be something similar to 
```cmd
./gradlew vendordep --install <url>
./gradlew vendordep --update
```
I know how to implement this for online install/update. I might need a bit of guidance with implementing offline.

### PR DESCRIPTION
**Pull request by Starlight220**
Fixes #417 
The original commit uses curl, the second uses the [gradle-download plugin](https://github.com/michel-kraemer/gradle-download-task).

### CODE REVIEW

File: src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/VendorDepTask.groovy
Comment (ThadHouse) [inline src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/VendorDepTask.groovy:11]: No need for this commented out line

File: src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/VendorDepTask.groovy
Comment (ThadHouse) [inline src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/VendorDepTask.groovy:44]: empty new line at end of file please.

File: src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/WPIDependenciesPlugin.groovy
Comment (ThadHouse) [inline src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/WPIDependenciesPlugin.groovy:55]: You dont need the as here. Instead, do like the below one does, and close the register block after the `VendorDepsTask` part. Groovy will then automatically take the closure afterwards as the 3rd parameter.

Comment (Starlight220) [reply src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/WPIDependenciesPlugin.groovy:55]: Fixed?

Comment (ThadHouse) [reply src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/WPIDependenciesPlugin.groovy:55]: Nope.

```
project.tasks.register("vendordep", VendorDepTask) { VendorDepTask task ->
            task.group = "GradleRIO"
            task.description = "Install vendordep JSON file from URL"
        }
```

Comment (Starlight220) [reply src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/WPIDependenciesPlugin.groovy:55]: So Kotlin got the last-lambda-parameter-as-block syntax from groovy... 
And is it fixed now?

**Comment by Starlight220** [comment]
How do I test the "feature" I added?

**Comment by ThadHouse** [comment]
Bump the version number, run the `publishToMavenLocal` task, and then use that updated version number in a normal robot project.

**Comment by Starlight220** [comment]
Don't merge this yet, it doesn't work.

**Comment by Starlight220** [comment]
Fixed.

**Comment by ThadHouse** [comment]
@Starlight220  looks good. Can you do a few things though. The first 2 before I'll merge it, the third just make a plan to do.

1. Update the download task to the latest version (its been updated)
2. Instead of just printing an error, also add an example of how to pass a url. Gradle doesn't print this by default, but for our users it'd be really helpful. Something like "No vendor JSON URL was entered. Try the following \n.\gradlew vendordep --url <insert_url_here>"
3. Work with @Daltz333 to get something added to docs for this.

**Comment by Starlight220** [comment]
3 is already done, it's waiting for this to be merged. The frc-docs PR is linked above. 
Will do 1,2 soon

**Comment by Starlight220** [comment]
@ThadHouse - I finished 1,2. Feel free to review/merge/whatever. 3 is wpilibsuite/frc-docs#755, feel free to check it out.

**Comment by Starlight220** [comment]
@ThadHouse - I was thinking about special-casing two strings in the task so that teams can use this task also to get the Command-Based JSONs with the task. maybe something like `gradlew vendordep --url=$COMMAND-NEW`. Preferably the first char would be an uncommon character ($ % & etc) so that it's a single char comparison to check if we have a URL or a special case (and if it's the latter we then check if the special case is valid etc). 
This would be helpful because right now there isn't a _conventional_ way to get the command-based JSONs. 

Either way, I think this feature would be more appropriate as a separate PR, assuming you think it should be added at all - what are your thoughts about this. All your reviews of this PR are fixed, can you review/merge it?

**Comment by ThadHouse** [comment]
How would that help. A user could just use the git url directly and pass it it. Definitely opposed to embedding that url in gradlerio, which is what it would take.

And there is a conventional way to get the command based json. VS Code, and theyre always installed with the installer.

**Comment by Starlight220** [comment]
I was more thinking about it fetching the JSON from the local wpi maven repo (where VS Code fetches it from anyway).

**Comment by Starlight220** [comment]
@ThadHouse should I add some docs about this to the `Adding Vendor Dependencies - Without VS Code` section in the `README.md`?

**Comment by ThadHouse** [comment]
Regarding the local url, if its just pointing to the local folder, I'm not opposed. Something like `$FRCLOCAL\filename.json` Just make sure its using starts with to make sure that only matches in the beginning.

**Comment by ThadHouse** [comment]
And yes, updating the readme would be helpful.

**Comment by Starlight220** [comment]
@ThadHouse how do I get the path to FRCHOME from plugin code?

**Comment by Starlight220** [comment]
I finally got the local fetching to work.
Needs some testing, I'll be testing these cases:
- [ ] 1. exists in FRCHOME, nonexistent in project, default vendor folder location {should add the file}

- [ ] 2. nonexistent in FRCHOME, nonexistent in project, default vendor folder location {should fail with `file not found` error message}

- [ ] 3. exists in FRCHOME, existent in project, default vendor folder location {should replace with warning/message}

- [ ] 4. same as (1) but non-default vendor folder location
- [ ] 5. same as (2) but non-default vendor folder location
- [ ] 6. same as (3) but non-default vendor folder location

- [ ] make sure `info` prints are correct

**Comment by Starlight220** [comment]
I now tested, the only issue I saw was that the overwriting warning isn't printed when fetching from remote.

**Review by ThadHouse — APPROVED** [review]

**Pull request merged**
